### ISSUE TITLE
Move the definition of the atomic operations to a "details" namespace ?

### ISSUE THREAD

**Issue by fwyzard** [alpaka-group/alpaka#1656]
The atomic operation types declared in `include/alpaka/atomic/Op.hpp` are in the `alpaka` namespace:
```c++
namespace alpaka
{
    //! The addition function object.
    struct AtomicAdd
    {
    ....
```

I would propose moving them to a nested private namespace, like `aplaka::detail`:
```c++
namespace alpaka::detail
{
    //! The addition function object.
    struct AtomicAdd
    {
    ....
```

I spent a couple of hours trying to debug why `alpaka::AtomicAdd(...)` was not compiling or doing what I wanted, before realising I should have written `alpaka::atomicAdd(...)` :-/

**Comment by psychocoderHPC** [comment]
I do not support this request.
I can understand your problem, in old alpaka released the operations were under the namespace `alpaka::op` but were moved later into the alpaka namespace.


The reason why I do not support the request is that the operations are part of the user interface. In other applications, we use `alpaka::atomicOp<Operation>(...)` or require the `AtomicYYY` signature to specialize atomic behaviors for user defined data types or implement our own atomic operation to workaround e.g slow atomicAdd operations into shared memory for our IO pattern when using ROCM https://github.com/ComputationalRadiationPhysics/picongpu/blob/02f6d46505a55244f0ee384f53b7353126ddec31/include/pmacc/kernel/atomic.hpp#L70-L126.

**Comment by psychocoderHPC** [comment]
I agree that the naming can be confusing and I am for moving the operations into an own namespace `op` or `operation` but not `detail`.

CCing: @bernhardmgruber @j-stephan

**Comment by fwyzard** [comment]
> In other applications, we use `alpaka::atomicOp<Operation>(...)` or require the `AtomicYYY` signature to specialize atomic behaviors for user defined data types or implement our own atomic operation to workaround e.g slow atomicAdd operations into shared memory for our IO pattern when using ROCM https://github.com/ComputationalRadiationPhysics/picongpu/blob/02f6d46505a55244f0ee384f53b7353126ddec31/include/pmacc/kernel/atomic.hpp#L70-L126.

Off topic in this issue, but wouldn't it make sense to have the workaround for the poor performance directly into Alpaka ?

**Comment by psychocoderHPC** [comment]
> Off topic in this issue, but wouldn't it make sense to have the workaround for the poor performance directly into Alpaka ?

I opened #1657

**Comment by bernhardmgruber** [comment]
> I spent a couple of hours trying to debug why alpaka::AtomicAdd(...) was not compiling or doing what I wanted, before realising I should have written alpaka::atomicAdd(...) :-/

@fwyzard I am trying to imagine in what kind of trap you fell wrt. the current API design. `alpaka::AtomicAdd(...)` with any non-zero argument list should not compile and you should get a reasonable error message here. What is the design bug here?

@psychocoderHPC I removed the `op` namespace back then to make atomic calls like `alpaka::atomicOp<alpaka::op::Add>(...)` less verbose. By now we have the direct shortcuts like `alpaka::atomicAdd(...)`, so I don't care so much anymore about the verbosity of getting the functors. If you want to really reintroduce the `op` namespace, then fine.

**Comment by fwyzard** [comment]
> @fwyzard I am trying to imagine in what kind of trap you fell wrt. the current API design. `alpaka::AtomicAdd(...)` with any non-zero argument list should not compile and you should get a reasonable error message here. What is the design bug here?

Indeed, it doesn't compile, but the error message is far from obvious:
```
backend-cuda.cc(45): error: no suitable user-defined conversion from "const alpaka::AccGpuCudaRt<alpaka_cuda::Dim1D, alpaka_cuda::Idx>" to "alpaka::AtomicAdd" exists
``` 

Just for the sake of argument, I've tried moving all `AtomicXxx` types into a nexted namespace, and the error became
```
backend-cuda.cc(45): error: namespace "alpaka" has no member "AtomicAdd"
```

which is at least somewhat clearer.
